### ISSUE TITLE
Link Contours to Spinning Loading Wheel

### ISSUE THREAD

**Issue by jamesdoyle21** [asascience/ott-coops#143]

**Comment by jamesdoyle21** [comment]
https://user-images.githubusercontent.com/82955395/137760601-8cb49299-77e2-4a56-a86d-e5f713f1ec31.mp4

In the video you can see that when a data layer is loading, the spinning loading wheel in the top right corner indicates is active until the data is fully loaded. For contours the loading wheel will trigger for a second and then stop. Contours can take awhile to load so it's important to have it synched right with the wheel

**Comment by mpiannucci** [comment]
For some reason when the user turns on contours, the reloading fo the layer is not picked up by the loading spinner (AKA the mapbox event API)

**Comment by ndellicarpini** [comment]
This is caused at `MapWmsLayer.tsx:135`, where the mapstate is set not loading if an error occurs while loading a layer. This is very common when loading any type of layer, the error attached seems to almost always happen at least once while loading a layer (especially when the eds server is struggling). This doesn't seem to affect the layer loading its full contents however, and so I'm wondering if having the loading state disabled on error makes sense for loading layers.

![Screenshot 2021-10-19 141202](https://user-images.githubusercontent.com/22432482/137967478-d97ada15-73a3-4c0e-b80d-7368dfffa9a3.png)

**Comment by mpiannucci** [comment]
Gotcha... so it fires an error if one tile fails for a given layer but it might just retry that layer. I would get rid of the error handling for now then I think? We cna revisit layer and try to find a more clever solution

**Comment by ndellicarpini** [comment]
Info dump incoming, appreciate any help on this. Branch: https://github.com/asascience/oceansmap-3/tree/nick-loading-fix 

So my initial statement that the loading spinner being prematurely stopped is caused by errors loading the tiles is half right. For every tile loaded, there are 2 callbacks: 1 on load start (`MapWmsLayer.tsx:103`) and 1 on load end (`MapWmsLayer.tsx:115` if successful, `MapWmsLayer.tsx:135` if error). The data sent to the callback contains an important field, `isSourceLoaded`, which is true if the entire source requested is loaded. In most cases, this value is only true on the callback for the last tile loaded, so it is easy to check when to stop the loading spinner. However, if an error occurs, `isSourceLoaded` is true. This would be a non-issue, as the loading spinner would just restart on the next tile load, but after the first error, every single tile is given the parameter `isSourceLoaded=true` for no reason that I could figure out. I think this is either caused by the backend, or some processing that the mapbox-gl library does that I couldn't pinpoint.

Now for some general weirdness I've noticed. When a layer is active, the loading and loaded callbacks are triggered for every single onHover for each catalog item in the Data Layers list (probably for the layer bounding box?). Although this isn't bad per se, it does require more code in the callbacks to filter out these interactions. Also, because these callbacks are only triggered on a layer load or an onHover, when you disable a layer before it is fully loaded, the spinner will continue indefinitely (at least while using the `isSourceLoaded` checks). An easy fix to this is to just have a `setLayerNotLoading` called with `removeSelectedCatalogItem`.

tldr: Whenever the first tile error occurs on tile load the entire layer is set to fully loaded.

**Comment by mpiannucci** [comment]
good detective work. So it seems like you have a good grasp on a potential fix for the second issue and you need to look into mapbox-gl API a little further to understand why it is doing what it is which I think is the best way to approach this.

**Comment by ndellicarpini** [comment]
I was actually able to fix the above issues. There is a function in mapbox-gl called `areTilesLoaded` that does not have the same issues when an error occurs like `isSourceLoaded`. This function allowed me to properly adjust the callbacks, although I'd appreciate any testing on this issue to verify.

PR Here: https://github.com/asascience/oceansmap-3/pull/103

**Comment by mpiannucci** [comment]
Wooo awesome job

**Comment by jamesdoyle21** [comment]
Niiice, I'll do some testing this afternoon and keep you posted

**Comment by mpiannucci** [comment]
Its not deployed yet james haha

**Comment by jamesdoyle21** [comment]
I need to test a few of the other issues from the last sprint as well. I'll hold off and do them all together. Just let me know when its deployed
